code/modules/reagents/Chemistry-Holder.dm

```diff
@@ -18,70 +20,40 @@
 	if(SSchemistry)
 		SSchemistry.active_holders -= src
 
-	for(var/datum/reagent/R in reagent_list)
-		qdel(R)
-	reagent_list.Cut()
-	reagent_list = null
+	LAZYCLEARLIST(reagent_data)
+	LAZYCLEARLIST(reagent_volumes)
 	if(my_atom && my_atom.reagents == src)
 		my_atom.reagents = null
 
 /* Internal procs */
 
 /datum/reagents/proc/apply_force(var/force) // applies force to the reagents inside it
-	for (var/datum/reagent/A in reagent_list)
-		A.apply_force(force)
+	for (var/_A in reagent_volumes)
+		var/decl/reagent/A = decls_repository.get_decl(_A)
+		A.apply_force(force, src)
 
-/datum/reagents/proc/get_free_space() // Returns free space.
-	return maximum_volume - total_volume
+/datum/reagents/proc/get_primary_reagent_name() // Returns the name of the reagent with the biggest volume.
+	var/decl/reagent/reagent = get_primary_reagent_decl()
+	if(reagent)
+		. = reagent.name
 
-/datum/reagents/proc/get_master_reagent() // Returns reference to the reagent with the biggest volume.
-	var/the_reagent = null
-	var/the_volume = 0
-
-	for(var/datum/reagent/A in reagent_list)
-		if(A.volume > the_volume)
-			the_volume = A.volume
-			the_reagent = A
-
-	return the_reagent
-
-/datum/reagents/proc/get_reagent(var/rtype) // Returns reference to reagent matching passed type
-	for(var/datum/reagent/A in reagent_list)
-		if (A.type == rtype)
-			return A
-
-	return null
-
-/datum/reagents/proc/get_master_reagent_name() // Returns the name of the reagent with the biggest volume.
-	var/the_name = null
-	var/the_volume = 0
-	for(var/datum/reagent/A in reagent_list)
-		if(A.volume > the_volume)
-			the_volume = A.volume
-			the_name = A.name
-
-	return the_name
-
-/datum/reagents/proc/get_master_reagent_type() // Returns the type of the reagent with the biggest volume.
-	var/the_type = null
-	var/the_volume = 0
-	for(var/datum/reagent/A in reagent_list)
-		if(A.volume > the_volume)
-			the_volume = A.volume
-			the_type = A.type
-
-	return the_type
+/datum/reagents/proc/get_primary_reagent_decl()
+	. = primary_reagent && decls_repository.get_decl(primary_reagent)
 
 /datum/reagents/proc/update_total() // Updates volume and temperature.
-
 	total_volume = 0
-
-	for(var/datum/reagent/R in reagent_list)
-		if(R.volume < MINIMUM_CHEMICAL_VOLUME)
-			del_reagent(R.type)
+	primary_reagent = null
+	for(var/R in reagent_volumes)
+		var/vol = reagent_volumes[R]
+		if(vol < MINIMUM_CHEMICAL_VOLUME)
+			LAZYREMOVE(reagent_volumes, R)
+			LAZYREMOVE(reagent_data, R)
```

Comment:
Neither can be null at this point, since you're iterating over the contents of `reagent_volumes` here.

Reply:
Technically, you can have a reagent without data, I think- wait, temperature, nevermind